os/arch/arm/src/s5j/sss/security_hal_wrapper.c

```diff
@@ -0,0 +1,1627 @@
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  This file is based on part of mbed TLS (https://tls.mbed.org)
+ */
+
+#define ciL    (sizeof(uint64_t))
+#define biL    (ciL << 3)               /* bits  in limb  */
+#define biH    (ciL << 2)               /* half limb size */
+#define HAL_ASN1_CHK_ADD(g, f) do { if ((ret = f) < 0) return HAL_FAIL; else   \
+	g += ret; } while (0)
+#define GET_BYTE( X, i )                                \
+	( ( ( X )->p[( i ) / ciL] >> ( ( ( i ) % ciL ) * 8 ) ) & 0xff )
+
+#define HAL_MPI_CHK(f) do { if ((ret = f) != 0) { ret = HAL_FAIL; goto cleanup; }} while (0)
+#define CHARS_TO_LIMBS(i) ( (i) / ciL + ( (i) % ciL != 0 ) )
+
+typedef struct _hal_mpi {
+	int s;
+	uint32_t n;
+	uint64_t *p;
+} hal_mpi;
+
+static void hal_mpi_zeroize(uint64_t *v, uint32_t n)
+{
+	volatile uint64_t *p = v; while( n-- ) *p++ = 0;
+}
+
+static void hal_mpi_init(hal_mpi *X)
+{
+	if (X == NULL) {
+		return;
+	}
+
+	X->s = 1;
+	X->n = 0;
+	X->p = NULL;
+}
+
+static void hal_mpi_free(hal_mpi *X)
+{
+	if (X == NULL) {
+		return;
+	}
+
+	if (X->p != NULL) {
+		hal_mpi_zeroize(X->p, X->n);
+		free(X->p);
+	}
+
+	X->s = 1;
+	X->n = 0;
+	X->p = NULL;
+}
+
+static int hal_mpi_grow(hal_mpi *X, uint32_t nblimbs)
+{
+	uint64_t *p;
+
+	if (nblimbs > 10000) {
+		return HAL_ALLOC_FAIL;
+	}
+
+	if (X->n < nblimbs)	{
+		if ((p = (uint64_t *)calloc(nblimbs, ciL)) == NULL) {
+			return HAL_ALLOC_FAIL;
+		}
+
+		if (X->p != NULL) {
+			memcpy(p, X->p, X->n * ciL);
+			hal_mpi_zeroize(X->p, X->n);
+			free(X->p);
+		}
+
+		X->n = nblimbs;
+		X->p = p;
+	}
+
+	return HAL_SUCCESS;
+}
+
+static int hal_mpi_lset(hal_mpi *X, int64_t z)
+{
+	int ret;
+
+	HAL_MPI_CHK(hal_mpi_grow(X, 1));
+	memset(X->p, 0, X->n * ciL);
+
+	X->p[0] = (z < 0) ? -z : z;
+	X->s    = (z < 0) ? -1 : 1;
+
+cleanup:
+	return ret;
+}
+
+
+static uint32_t hal_clz(const uint64_t x)
+{
+	uint32_t j;
+	uint64_t mask = (uint64_t)1 << (biL - 1);
+
+	for (j = 0; j < biL; j++) {
+		if (x & mask) {
+			break;
+		}
+		mask >>= 1;
+	}
+
+	return j;
+}
+
+static uint32_t hal_mpi_bitlen(const hal_mpi *X)
+{
+	uint32_t i;
+	uint32_t j;
+
+	if (X->n == 0) {
+		return 0;
+	}
+
+	for (i = X->n - 1; i > 0; i--) {
+		if (X->p[i] != 0) {
+			break;
+		}
+	}
+	j = biL - hal_clz(X->p[i]);
+
+	return (i * biL) + j;
+}
+
+static uint32_t hal_mpi_size(const hal_mpi *X)
+{
+	return (hal_mpi_bitlen(X) + 7) >> 3;
+}
+
+static int hal_asn1_get_len(unsigned char **p, const unsigned char *end, uint32_t *len)
+{
+	if ((end - *p) < 1) {
+		return HAL_INVALID_ARGS;
+	}
+
+	if ((**p & 0x80) == 0) {
+		*len = *(*p)++;
+	} else {
+		switch (**p & 0x7F) {
+			case 1:
```

Comment:
Think you can clean this by making 1 simple loop.

Reply:
This format is standard ASN1 encoding param provided by mbedtls.